### PR TITLE
user_guide: Add section about 'Key Learning to Verify Files'

### PR DESCRIPTION
**Pull request by stefanberger**

### CODE REVIEW

**Comment by stefanberger** [comment]
We can only merge this once Fedora 38 can be used to follow these steps. At the moment it's not possible due to some bugs.
I am tracking these bugs via BZs:

    https://bugzilla.redhat.com/show_bug.cgi?id=2231396
    https://bugzilla.redhat.com/show_bug.cgi?id=2231404

Also see: https://github.com/keylime/enhancements/pull/52#issuecomment-1674104354

File: docs/user_guide/runtime_ima.rst [hunk 1]
```diff
@@ -461,6 +461,124 @@ The verifier log should now indicating a bad file signature::
...
+
+   dnf -y reinstall \*
+
+To verify whether a particular file has its file signature installed use
+the following command to display the contents of :code:`security.ima`. If
+nothing is displayed then this file misses its file signature::
+
+   getfattr -m ^securty.ima -e hex --dump /usr/bin/bash
+
+We must setup the system with the Linux boot command line option
```
Comment (aplanas) [inline docs/user_guide/runtime_ima.rst:494]: I saw it referenced more frequently as "the kernel command line"

File: docs/user_guide/runtime_ima.rst [hunk 2]
```diff
@@ -461,6 +461,124 @@ The verifier log should now indicating a bad file signature::
...
+this we edit :code:`/etc/default/grub` and adjust the following line::
+
+   GRUB_CMDLINE_LINUX="rhgb quiet ima_template=ima-sig"
+
+The run the following command to update the Linux boot command line options
+for the system::
+
+   grub2-mkconfig -o /boot/grub2/grub.conf
+
+Set the following IMA policy in :code:`/etc/ima/ima-policy`::
```
[the remark this answers is not in the document]
Comment (stefanberger) [reply docs/user_guide/runtime_ima.rst]: Done. Thanks. I will repeat the exercise (today) with F39.

**Comment by stefanberger** [comment]
FYI: Fedora 39 does not and cannot load the IMA keys onto the .ima keyring, yet.

With Fedora 39 I currently use the following commands:
```
grep -E "(boot_aggregate| ima-buf |keylime|livesys)" \
  /sys/kernel/security/ima/ascii_runtime_measurements > trimmed_ima_log
keylime_create_policy \
  -k \
  -m ./trimmed_ima_log \
  -o mypolicy.json \
  -A /etc/keys/ima/fedora-39-ima.der \
  -A /etc/keys/ima/fedora-38-ima.der

 keylime_tenant -c update --uuid <agent-uuid> -f payload --runtime-policy ./mypolicy.json
```

**Comment by stefanberger** [comment]
FYI: CentOS is now farther ahead with being able to load the IMA key into the kernel:

```
keyctl padd asymmetric "" %keyring:.ima < /usr/share/doc/kernel-keys/5.14.0-361.el9.x86_64/ima.cer
```

But it's farther behind on signatures -- some packages have signatures missing (maybe were built before signatures were enabled?):

```
[root@localhost ~]# rpm -q --whatprovides /usr/bin/find
findutils-4.8.0-6.el9.x86_64
[root@localhost ~]# getfattr -m ^security -e hex --dump /usr/bin/find
getfattr: Removing leading '/' from absolute path names
# file: usr/bin/find
security.selinux=0x73797374656d5f753a6f626a6563745f723a62696e5f743a733000
```

**Comment by stefanberger** [comment]
FYI: On CentOS some of the length indicators of file signatures have a messed up length indicator:
```
# getfattr -m ^security.ima -e hex --dump /usr/bin/bash
getfattr: Removing leading '/' from absolute path names
# file: usr/bin/bash
security.ima=0x0302047c435bb500673065023061c39ffe4418f278d720ed0e6f07b9c0e83a8d0c4bc69529747dd10ab31adf9b734874326c300208cd9a35b19ce5f01c0231008d198181be551287334e4ae1ca2ac61cd96b94e486ed64617b8e2f24582fd89e4ce7634a0e07215fb3b51925a9b8fe5a

# getfattr -m ^security.ima -e hex --dump /usr/lib64/libacl.so.1.1.2301
getfattr: Removing leading '/' from absolute path names
# file: usr/lib64/libacl.so.1.1.2301
security.ima=0x0302047c435bb56600306402302358e7ffb330e36d7d4b999d379aaefaac1fa4039cdbd101a4bfcc246910eb367a90d180e95de8c805910124c9dcc054023059d6c0dcff551c4ac32573cf796289f7b9c95782223a400291b0cd17985530e094d51be10ae88234d1769f9222f31a03

```

Compare the two where libacl has a little endian signature length indicator:
```
security.ima=0x0302047c435bb50067   -> 0067
security.ima=0x0302047c435bb56600   -> 6600
```

File: docs/user_guide/runtime_ima.rst [hunk 3]
```diff
@@ -461,6 +461,143 @@ The verifier log should now indicating a bad file signature::
...
+
+We must setup the system with the kernel command line option
+:code:`ima_template=ima-sig` so that IMA signatures become part of the measurement
+log. It is not necessary to enable signature enforcement on the system, measuring
+executed applications is sufficient for the purpose of 'key learning'. For
+this we edit :code:`/etc/default/grub` and adjust the following line::
+
+   GRUB_CMDLINE_LINUX="rhgb quiet ima_template=ima-sig"
+
+The run the following command to update the kernel command line options::
```
Comment (stefanberger) [inline docs/user_guide/runtime_ima.rst:511]: s/The/Then

File: docs/user_guide/runtime_ima.rst [hunk 4]
```diff
@@ -461,6 +461,143 @@ The verifier log should now indicating a bad file signature::
...
+     /sys/kernel/security/ima/ascii_runtime_measurements > trimmed_ima_log
+
+   keylime_create_policy -k -m ./trimmed_ima_log -o mypolicy.json
+
+The 1st command creates a trimmed-down IMA measurement log that only
+contains the boot_aggregate and ima-buf entries. The latter show the key(s)
+that were loaded onto the :code:`.ima` keyring.
+
+The 2nd command creates the runtime policy that holds the boot_aggregate
+entry and a hash over the key that was loaded onto the .ima keyring. This
```
Comment (stefanberger) [inline docs/user_guide/runtime_ima.rst:572]: key ->key(s)

File: docs/user_guide/runtime_ima.rst [hunk 5]
```diff
@@ -461,6 +461,143 @@ The verifier log should now indicating a bad file signature::
   keylime.cloudverifier - WARNING - agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000 failed, stopping polling
 
 
+Using Key Learning to Verify Files
+----------------------------------
+
+Using key learning to verify files requires that files logged by IMA are
+appropriately signed. If files are not signed or have a bad signature then
+they must be either in the exclude list of the runtime policy or their hashes
+must be part of the runtime policy. It should be noted also that IMA signature
```
Comment (stefanberger) [inline docs/user_guide/runtime_ima.rst:470]: also be noted

File: docs/user_guide/runtime_ima.rst [hunk 6]
```diff
@@ -461,6 +461,143 @@ The verifier log should now indicating a bad file signature::
...
+verification provides lock-down of a system and ensures the provenance of
+files but does not provide the protection for renaming of files as well
```
Comment (stefanberger) [inline docs/user_guide/runtime_ima.rst:472]: does not provide protection

File: docs/user_guide/runtime_ima.rst [hunk 7]
```diff
@@ -461,6 +461,143 @@ The verifier log should now indicating a bad file signature::
...
+as replacing files and signatures with other versions (downgrading).
```
Comment (stefanberger) [inline docs/user_guide/runtime_ima.rst:473]: , unlike file hashes.

**Comment by THS-on** [comment]
@stefanberger what is the upstream status of this? Do we have a distro where those steps work out-of-the-box?

**Comment by stefanberger** [comment]
Only RHEL works but not (yet) the ones described in the document, CentOS and Fedora.

**Comment by THS-on** [comment]
What do you think about changing it to RHEL and then adding a comment that it currently is still WIP for Fedora and CentOS?

**Comment by stefanberger** [comment]
> What do you think about changing it to RHEL and then adding a comment that it currently is still WIP for Fedora and CentOS?

I tested it now with RHEL 9.3 and adjusted the text.

**Review by THS-on — APPROVED** [review]

**Pull request merged**